steward/src/cellars/adaptors/balancer_pool.rs

```diff
@@ -0,0 +1,174 @@
+                };
+                calls.push(
+                    BalancerPoolAdaptorV1Calls::RevokeApproval(call)
+                        .encode()
+                        .into(),
+                )
+            }
+            balancer_pool_adaptor_v1::Function::JoinPool(p) => {
+                if !p.swaps_before_join.is_empty()
+                    && p.swaps_before_join.iter().any(|s| s.kind == 0)
+                {
+                    return Err(sp_call_error("invalid swap kind".to_string()));
+                }
+
+                let swaps_before_join = p
+                    .swaps_before_join
+                    .into_iter()
+                    .map(|s| {
+                        Ok(steward_abi::balancer_pool_adaptor_v1::SingleSwap {
+                            pool_id: sp_call_parse_h256(s.pool_id)?.into(),
+                            kind: (s.kind - 1) as u8,
+                            asset_in: sp_call_parse_address(s.asset_in)?,
+                            asset_out: sp_call_parse_address(s.asset_out)?,
+                            amount: string_to_u256(s.amount)?,
+                            user_data: s.user_data.into(),
+                        })
+                    })
+                    .collect::<Result<Vec<_>, Error>>()?;
+
+                let swap_data = match p.swap_data {
+                    Some(d) => SwapData {
+                        min_amounts_for_swaps: d
+                            .min_amounts_for_swaps
+                            .into_iter()
+                            .map(string_to_u256)
+                            .collect::<Result<Vec<_>, Error>>()?,
+                        swap_deadlines: d
+                            .swap_deadlines
+                            .into_iter()
+                            .map(string_to_u256)
+                            .collect::<Result<Vec<_>, Error>>()?,
+                    },
+                    None => return Err(sp_call_error("swap data must be set".to_string())),
+                };
+
+                let call = steward_abi::balancer_pool_adaptor_v1::JoinPoolCall {
+                    target_bpt: sp_call_parse_address(p.target_bpt)?,
+                    swaps_before_join,
+                    swap_data,
+                    minimum_bpt: string_to_u256(p.minimum_bpt)?,
+                };
+
+                calls.push(BalancerPoolAdaptorV1Calls::JoinPool(call).encode().into())
+            }
+            balancer_pool_adaptor_v1::Function::ExitPool(p) => {
+                if !p.swaps_after_exit.is_empty() && p.swaps_after_exit.iter().any(|s| s.kind == 0)
+                {
+                    return Err(sp_call_error("invalid swap kind".to_string()));
+                }
+
+                let swaps_after_exit = p
+                    .swaps_after_exit
+                    .into_iter()
+                    .map(|s| {
+                        Ok(steward_abi::balancer_pool_adaptor_v1::SingleSwap {
```

Comment:
Since we are encoding `SingleSwap` and `SwapData` a bunch of times, we may want to break this out into its own function.